In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

# Для работы с матрицами
from scipy.sparse import csr_matrix, coo_matrix

# Матричная факторизация
#from implicit.als import AlternatingLeastSquares
#from implicit.nearest_neighbours import bm25_weight, tfidf_weight

from lightfm import LightFM

# Функции из 1-ого вебинара
import os, sys

module_path = os.path.abspath(os.path.join(os.pardir))
if module_path not in sys.path:
    sys.path.append(module_path)

/opt/anaconda3/lib/python3.9/site-packages/lightfm/_lightfm_fast.py:9: UserWarning: LightFM was compiled without OpenMP support. Only a single thread will be used.
  warnings.warn(


In [2]:
!pip install lightfm

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 310.1/310.1 kB 1.8 MB/s eta 0:00:00a 0:00:01
  Preparing metadata (setup.py) ... done
  Created wheel for lightfm: filename=lightfm-1.16-cp39-cp39-macosx_10_9_x86_64.whl size=425634 sha256=d6ffd6eeb38f0cbb31ef6915a30d1576f378261d2349dde3dd575f8919ea38ee
  Stored in directory: /Users/olgapodosenova/Library/Caches/pip/wheels/b5/0d/a5/aed8bf2edc20773ee85f7eeb108d6babf2732675c0d6170aaa
Successfully built lightfm


In [4]:
from lightfm.evaluation import precision_at_k, recall_at_k

from metrics import precision_at_k as custom_precision, recall_at_k
from utils import prefilter_items

In [5]:
data = pd.read_csv('retail_train.csv')

item_features = pd.read_csv('product.csv')
user_features = pd.read_csv('hh_demographic.csv')

# column processing
item_features.columns = [col.lower() for col in item_features.columns]
user_features.columns = [col.lower() for col in user_features.columns]

item_features.rename(columns={'product_id': 'item_id'}, inplace=True)
user_features.rename(columns={'household_key': 'user_id'}, inplace=True)

# train test split
test_size_weeks = 3

data_train = data[data['week_no'] < data['week_no'].max() - test_size_weeks]
data_test = data[data['week_no'] >= data['week_no'].max() - test_size_weeks]

data_train.head(2)

user_id    basket_id  day  item_id  quantity  sales_value  store_id  \
0     2375  26984851472    1  1004906         1         1.39       364   
1     2375  26984851472    1  1033142         1         0.82       364   

   retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  
0         -0.6        1631        1          0.0                0.0  
1          0.0        1631        1          0.0                0.0

In [6]:
result = data_test.groupby('user_id')['item_id'].unique().reset_index()
result.columns=['user_id', 'actual']
result.head(2)

user_id                                             actual
0        1  [821867, 834484, 856942, 865456, 889248, 90795...
1        3  [835476, 851057, 872021, 878302, 879948, 90963...

In [7]:
item_features.head(2)

item_id  manufacturer    department     brand            commodity_desc  \
0    25671             2       GROCERY  National                  FRZN ICE   
1    26081             2  MISC. TRANS.  National  NO COMMODITY DESCRIPTION   

            sub_commodity_desc curr_size_of_product  
0          ICE - CRUSHED/CUBED                22 LB  
1  NO SUBCOMMODITY DESCRIPTION

In [8]:
user_features.head(2)

age_desc marital_status_code income_desc homeowner_desc      hh_comp_desc  \
0      65+                   A      35-49K      Homeowner  2 Adults No Kids   
1    45-54                   A      50-74K      Homeowner  2 Adults No Kids   

  household_size_desc kid_category_desc  user_id  
0                   2      None/Unknown        1  
1                   2      None/Unknown        7

In [9]:
user_features['age_desc'].unique()

array(['65+', '45-54', '25-34', '35-44', '19-24', '55-64'], dtype=object)

In [10]:
user_features['marital_status_code'].unique()

array(['A', 'U', 'B'], dtype=object)

In [11]:
user_features['household_size_desc'].unique()

array(['2', '3', '4', '1', '5+'], dtype=object)

1. Filter items

In [12]:
data_train.head()

user_id    basket_id  day  item_id  quantity  sales_value  store_id  \
0     2375  26984851472    1  1004906         1         1.39       364   
1     2375  26984851472    1  1033142         1         0.82       364   
2     2375  26984851472    1  1036325         1         0.99       364   
3     2375  26984851472    1  1082185         1         1.21       364   
4     2375  26984851472    1  8160430         1         1.50       364   

   retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  
0        -0.60        1631        1          0.0                0.0  
1         0.00        1631        1          0.0                0.0  
2        -0.30        1631        1          0.0                0.0  
3         0.00        1631        1          0.0                0.0  
4        -0.39        1631        1          0.0                0.0

In [13]:
n_items_before = data_train['item_id'].nunique()

data_train_filtered = prefilter_items(data_train, take_n_popular=5000, item_features=item_features)

n_items_after = data_train_filtered['item_id'].nunique()
print('Decreased # items from {} to {}'.format(n_items_before, n_items_after))

Decreased # items from 86865 to 5001


/Users/olgapodosenova/RS/utils.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['price'] = data['sales_value'] / (np.maximum(data['quantity'], 1))


In [14]:
data_train_filtered['item_id'].value_counts()

999999      303002
1029743      10378
1106523       7185
5569230       4134
916122        3893
             ...  
5570974         13
859809          12
898448          10
12696099        10
1058629          7
Name: item_id, Length: 5001, dtype: int64

2. Prepare data set

2.1 Prepare csr train matrix

In [15]:
user_item_matrix = pd.pivot_table(data_train_filtered, 
                                  index='user_id', columns='item_id', 
                                  values='quantity', # Можно пробовать другие варианты
                                  aggfunc='count', 
                                  fill_value=0
                                 )

user_item_matrix = user_item_matrix.astype(float) # необходимый тип матрицы для implicit

# переведем в формат sparse matrix
sparse_user_item = csr_matrix(user_item_matrix).tocsr()

user_item_matrix.head(2)

item_id  117847    818981    819255    819308    819400    819487    819590    \
user_id                                                                         
1             0.0       0.0       0.0       0.0       0.0       0.0       0.0   
2             0.0       0.0       0.0       0.0       0.0       0.0       0.0   

item_id  819594    819840    819845    ...  15926775  15926844  15926886  \
user_id                                ...                                 
1             0.0       0.0       0.0  ...       0.0       1.0       0.0   
2             0.0       0.0       0.0  ...       0.0       0.0       0.0   

item_id  15972074  15972298  15972565  15972790  16100266  16729299  16729415  
user_id                                                                        
1             0.0       0.0       0.0       0.0       0.0       0.0       0.0  
2             0.0       0.0       0.0       0.0       0.0       0.0       0.0  

[2 rows x 5001 columns]

2.2 Prepare CSR test matrix

In [16]:
data_test = data_test[data_test['item_id'].isin(data_train['item_id'].unique())]

test_user_item_matrix = pd.pivot_table(data_test, 
                                  index='user_id', columns='item_id', 
                                  values='quantity', # Можно пробовать другие варианты
                                  aggfunc='count', 
                                  fill_value=0
                                 )

test_user_item_matrix = test_user_item_matrix.astype(float) # необходимый тип матрицы для implicit

In [17]:
userids = user_item_matrix.index.values
itemids = user_item_matrix.columns.values

matrix_userids = np.arange(len(userids))
matrix_itemids = np.arange(len(itemids))

id_to_itemid = dict(zip(matrix_itemids, itemids))
id_to_userid = dict(zip(matrix_userids, userids))

itemid_to_id = dict(zip(itemids, matrix_itemids))
userid_to_id = dict(zip(userids, matrix_userids))

3. Prepare user and item features

In [18]:
user_feat = pd.DataFrame(user_item_matrix.index)
user_feat = user_feat.merge(user_features, on='user_id', how='left')
user_feat.set_index('user_id', inplace=True)
user_feat.head(2)

age_desc marital_status_code income_desc homeowner_desc  \
user_id                                                           
1            65+                   A      35-49K      Homeowner   
2            NaN                 NaN         NaN            NaN   

             hh_comp_desc household_size_desc kid_category_desc  
user_id                                                          
1        2 Adults No Kids                   2      None/Unknown  
2                     NaN                 NaN               NaN

In [19]:
user_feat.shape

(2497, 7)

In [20]:
item_feat = pd.DataFrame(user_item_matrix.columns)
item_feat = item_feat.merge(item_features, on='item_id', how='left')
item_feat.set_index('item_id', inplace=True)

item_feat.head(2)

manufacturer department     brand commodity_desc sub_commodity_desc  \
item_id                                                                        
117847          450.0  NUTRITION  National   REFRIGERATED      SOY/RICE MILK   
818981          194.0    GROCERY  National    COLD CEREAL  ALL FAMILY CEREAL   

        curr_size_of_product  
item_id                       
117847                 64 OZ  
818981               10.4 OZ

In [21]:
item_feat.shape

(5001, 6)

Encoding features

In [22]:
user_feat_lightfm = pd.get_dummies(user_feat, columns=user_feat.columns.tolist())
item_feat_lightfm = pd.get_dummies(item_feat, columns=item_feat.columns.tolist())

In [23]:
user_feat_lightfm.head(2)

age_desc_19-24  age_desc_25-34  age_desc_35-44  age_desc_45-54  \
user_id                                                                   
1                     0               0               0               0   
2                     0               0               0               0   

         age_desc_55-64  age_desc_65+  marital_status_code_A  \
user_id                                                        
1                     0             1                      1   
2                     0             0                      0   

         marital_status_code_B  marital_status_code_U  income_desc_100-124K  \
user_id                                                                       
1                            0                      0                     0   
2                            0                      0                     0   

         ...  hh_comp_desc_Unknown  household_size_desc_1  \
user_id  ...                                                
1        ...                     0                      0   
2        ...                     0                      0   

         household_size_desc_2  household_size_desc_3  household_size_desc_4  \
user_id                                                                        
1                            1                      0                      0   
2                            0                      0                      0   

         household_size_desc_5+  kid_category_desc_1  kid_category_desc_2  \
user_id                                                                     
1                             0                    0                    0   
2                             0                    0                    0   

         kid_category_desc_3+  kid_category_desc_None/Unknown  
user_id                                                        
1                           0                               1  
2                           0                               0  

[2 rows x 41 columns]

Init model

In [24]:
model = LightFM(no_components=40,
                loss='bpr', # "logistic","bpr", "warp"
                learning_rate=0.01, 
                item_alpha=0.4,
                user_alpha=0.1, 
                random_state=42,
                k=5,
                n=15,
                max_sampled=100)

Train

In [25]:
model.fit((sparse_user_item > 0) * 1,  # user-item matrix из 0 и 1
          sample_weight=coo_matrix(user_item_matrix),
          user_features=csr_matrix(user_feat_lightfm.values).tocsr(),
          item_features=csr_matrix(item_feat_lightfm.values).tocsr(),
          epochs=20, 
          num_threads=20,
          verbose=True) 

Epoch: 100%|████████████████████████████████████| 20/20 [00:12<00:00,  1.63it/s]


Predict

In [26]:
# подготавливаемм id для юзеров и товаров в порядке пар user-item
users_ids_row = data_train_filtered['user_id'].apply(lambda x: userid_to_id[x]).values.astype(int)
items_ids_row = data_train_filtered['item_id'].apply(lambda x: itemid_to_id[x]).values.astype(int)

In [27]:
users_ids_row[:10]

array([2371, 1363, 1363, 1363, 1363, 1171, 1171, 1171, 1171, 1171])

In [28]:
# модель возвращает меру/скор похожести между соответствующим пользователем и товаром
predictions = model.predict(user_ids=users_ids_row,
                            item_ids=items_ids_row,
                            user_features=csr_matrix(user_feat_lightfm.values).tocsr(),
                            item_features=csr_matrix(item_feat_lightfm.values).tocsr(),
                            num_threads=10)

In [29]:
# добавляем наш полученный скор в трейн датафрейм
data_train_filtered['score'] = predictions

In [30]:
data_train_filtered.head()

user_id    basket_id  day  item_id  quantity  sales_value  store_id  \
7      2375  26984851516    1  1085983         1         2.99       364   
11     1364  26984896261    1   999999         1         2.19     31742   
12     1364  26984896261    1   999999         1         2.99     31742   
13     1364  26984896261    1   999999         1         3.09     31742   
14     1364  26984896261    1   937406         1         2.50     31742   

    retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  price  \
7         -0.40        1642        1          0.0                0.0   2.99   
11         0.00        1520        1          0.0                0.0   2.19   
12        -0.40        1520        1          0.0                0.0   2.99   
13         0.00        1520        1          0.0                0.0   3.09   
14        -0.99        1520        1          0.0                0.0   2.50   

       score  
7  -0.023003  
11 -0.983615  
12 -0.983615  
13 -0.983615  
14 -1.003001

In [31]:
# создаем предикт датафрейм в формате списка това
predict_result = data_train_filtered[['user_id','item_id','score']][data_train_filtered.item_id != 999999].drop_duplicates().sort_values(by=['user_id','score'], ascending=False).groupby('user_id')['item_id']. \
            unique().reset_index()

In [32]:
predict_result.head()

user_id                                            item_id
0        1  [1029743, 6034857, 952163, 1004906, 1088462, 8...
1        2  [1106523, 1075368, 6919458, 952163, 899624, 83...
2        3  [1106523, 983584, 5585510, 899624, 946839, 839...
3        4  [1029743, 1075368, 1052294, 7431408, 970760, 1...
4        5  [1126899, 1029743, 6034991, 825659, 1112387, 1...

In [33]:
# объединяем предикт и тест датасет для подсчета precision
df_result_for_metrics = result.merge(predict_result, on='user_id', how='inner')

In [34]:
df_result_for_metrics.head()

user_id                                             actual  \
0        1  [821867, 834484, 856942, 865456, 889248, 90795...   
1        3  [835476, 851057, 872021, 878302, 879948, 90963...   
2        6  [920308, 926804, 946489, 1006718, 1017061, 107...   
3        7  [840386, 889774, 898068, 909714, 929067, 95347...   
4        8  [835098, 872137, 910439, 924610, 992977, 10412...   

                                             item_id  
0  [1029743, 6034857, 952163, 1004906, 1088462, 8...  
1  [1106523, 983584, 5585510, 899624, 946839, 839...  
2  [1070820, 1029743, 1126899, 1121393, 9524291, ...  
3  [1029743, 1126899, 1106523, 1072494, 9835880, ...  
4  [1106523, 1070820, 1029743, 5585510, 1139835, ...

Test with custom precision func

In [35]:
precision = df_result_for_metrics.apply(lambda row: custom_precision(row['item_id'], row['actual'],k=5), axis=1).mean()
print(f"Precision: {precision}")

Precision: 0.13963875205254345


Домашнее задание
1) Прочитать статьи про BPR, WARP loss

2) Сделать грид серч текущей модели, смотрите на метрику precision@5, считаем на тесте нашей функцией

Подбор параметров:
- Loss
- no_components
- regularization

In [36]:
n_comp = [40, 80, 150]
losses = ["bpr", "warp"] 
learning_rates = [0.1, 0.01]
item_alphas = [0.5, 0.3]
user_alphas = [0.1, 0.05]
max_samples = [100, 150]
precisions = {
    'n_comp' : [],
    'losses' : [],
    'learning_rates' : [],
    'item_alphas' : [],
    'user_alphas' : [],
    'max_samples' : [],
    'precisions' : []
}

In [37]:
for n_c in n_comp:
  for l in losses:
    for lr in learning_rates:
      for ia in item_alphas:
        for ua in user_alphas:
          for ms in max_samples:

            model = LightFM(no_components=n_c,
                            loss=l, # "bpr", "warp"
                            learning_rate=lr, 
                            item_alpha=ia,
                            user_alpha=ua, 
                            random_state=42,
                            k=5,
                            n=15,
                            max_sampled=ms)
            model.fit((sparse_user_item > 0) * 1,  # user-item matrix из 0 и 1
                      sample_weight=coo_matrix(user_item_matrix),
                      user_features=csr_matrix(user_feat_lightfm.values).tocsr(),
                      item_features=csr_matrix(item_feat_lightfm.values).tocsr(),
                      epochs=10, 
                      num_threads=20,
                      verbose=True) 
            users_ids_row = data_train_filtered['user_id'].apply(lambda x: userid_to_id[x]).values.astype(int)
            items_ids_row = data_train_filtered['item_id'].apply(lambda x: itemid_to_id[x]).values.astype(int)

            predictions = model.predict(user_ids=users_ids_row,
                                        item_ids=items_ids_row,
                                        user_features=csr_matrix(user_feat_lightfm.values).tocsr(),
                                        item_features=csr_matrix(item_feat_lightfm.values).tocsr(),
                                        num_threads=10)

            data_train_filtered['score'] = predictions

            predict_result = data_train_filtered[['user_id','item_id','score']][data_train_filtered.item_id != 999999].drop_duplicates().sort_values(by=['user_id','score'], ascending=False).groupby('user_id')['item_id']. \
                        unique().reset_index()

            df_result_for_metrics = result.merge(predict_result, on='user_id', how='inner')
            precision = df_result_for_metrics.apply(lambda row: custom_precision(row['item_id'], row['actual'],k=5), axis=1).mean()


            precisions['n_comp'].append(n_c)
            precisions['losses'].append(l) 
            precisions['learning_rates'].append(lr) 
            precisions['item_alphas'].append(ia) 
            precisions['user_alphas'].append(ua) 
            precisions['max_samples'].append(ms)
            precisions['precisions'].append(precision)
            print(len(precisions['precisions']), 'Из 96')

Epoch: 100%|████████████████████████████████████| 10/10 [00:06<00:00,  1.64it/s]


1 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:06<00:00,  1.61it/s]


2 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:06<00:00,  1.63it/s]


3 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:06<00:00,  1.61it/s]


4 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:06<00:00,  1.62it/s]


5 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:06<00:00,  1.61it/s]


6 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:06<00:00,  1.62it/s]


7 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:06<00:00,  1.61it/s]


8 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:06<00:00,  1.62it/s]


9 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:06<00:00,  1.61it/s]


10 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:06<00:00,  1.62it/s]


11 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:06<00:00,  1.61it/s]


12 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:06<00:00,  1.62it/s]


13 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:06<00:00,  1.62it/s]


14 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:06<00:00,  1.59it/s]


15 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:06<00:00,  1.62it/s]


16 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:07<00:00,  1.31it/s]


17 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:07<00:00,  1.26it/s]


18 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:07<00:00,  1.28it/s]


19 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:07<00:00,  1.25it/s]


20 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:07<00:00,  1.31it/s]


21 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:08<00:00,  1.25it/s]


22 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:07<00:00,  1.29it/s]


23 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:08<00:00,  1.22it/s]


24 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:08<00:00,  1.24it/s]


25 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:08<00:00,  1.15it/s]


26 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:08<00:00,  1.24it/s]


27 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:08<00:00,  1.16it/s]


28 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:07<00:00,  1.30it/s]


29 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:08<00:00,  1.25it/s]


30 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:08<00:00,  1.23it/s]


31 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:08<00:00,  1.16it/s]


32 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:11<00:00,  1.14s/it]


33 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:11<00:00,  1.14s/it]


34 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:11<00:00,  1.14s/it]


35 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:11<00:00,  1.16s/it]


36 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:11<00:00,  1.15s/it]


37 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:11<00:00,  1.14s/it]


38 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:11<00:00,  1.14s/it]


39 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:11<00:00,  1.15s/it]


40 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:11<00:00,  1.14s/it]


41 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:11<00:00,  1.17s/it]


42 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:11<00:00,  1.17s/it]


43 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:11<00:00,  1.17s/it]


44 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:11<00:00,  1.14s/it]


45 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:11<00:00,  1.14s/it]


46 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:11<00:00,  1.15s/it]


47 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:11<00:00,  1.17s/it]


48 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:13<00:00,  1.33s/it]


49 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:13<00:00,  1.38s/it]


50 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:13<00:00,  1.34s/it]


51 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:14<00:00,  1.40s/it]


52 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:14<00:00,  1.48s/it]


53 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:14<00:00,  1.49s/it]


54 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:14<00:00,  1.42s/it]


55 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:14<00:00,  1.43s/it]


56 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:14<00:00,  1.43s/it]


57 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:15<00:00,  1.59s/it]


58 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:14<00:00,  1.50s/it]


59 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:15<00:00,  1.54s/it]


60 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:13<00:00,  1.34s/it]


61 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:13<00:00,  1.37s/it]


62 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:14<00:00,  1.42s/it]


63 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:15<00:00,  1.53s/it]


64 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:22<00:00,  2.24s/it]


65 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:21<00:00,  2.19s/it]


66 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:21<00:00,  2.16s/it]


67 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:21<00:00,  2.15s/it]


68 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:20<00:00,  2.09s/it]


69 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:21<00:00,  2.12s/it]


70 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:21<00:00,  2.16s/it]


71 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:21<00:00,  2.16s/it]


72 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:20<00:00,  2.07s/it]


73 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:21<00:00,  2.11s/it]


74 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:20<00:00,  2.09s/it]


75 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:21<00:00,  2.16s/it]


76 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:21<00:00,  2.16s/it]


77 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:21<00:00,  2.16s/it]


78 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:21<00:00,  2.15s/it]


79 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:21<00:00,  2.11s/it]


80 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:24<00:00,  2.49s/it]


81 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:25<00:00,  2.58s/it]


82 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:24<00:00,  2.48s/it]


83 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:25<00:00,  2.58s/it]


84 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:25<00:00,  2.52s/it]


85 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:32<00:00,  3.24s/it]


86 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:23<00:00,  2.40s/it]


87 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:25<00:00,  2.54s/it]


88 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:25<00:00,  2.52s/it]


89 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:26<00:00,  2.62s/it]


90 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:25<00:00,  2.51s/it]


91 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:26<00:00,  2.69s/it]


92 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:24<00:00,  2.41s/it]


93 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:23<00:00,  2.32s/it]


94 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:24<00:00,  2.47s/it]


95 Из 96


Epoch: 100%|████████████████████████████████████| 10/10 [00:25<00:00,  2.58s/it]


96 Из 96


In [38]:
df = pd.DataFrame(precisions)

In [39]:
df = df.sort_values('precisions', ascending=False)

In [40]:
df

n_comp losses  learning_rates  item_alphas  user_alphas  max_samples  \
89     150   warp            0.01          0.5         0.10          150   
61      80   warp            0.01          0.3         0.10          150   
14      40    bpr            0.01          0.3         0.05          100   
15      40    bpr            0.01          0.3         0.05          150   
92     150   warp            0.01          0.3         0.10          100   
..     ...    ...             ...          ...          ...          ...   
70     150    bpr            0.10          0.3         0.05          100   
69     150    bpr            0.10          0.3         0.10          150   
68     150    bpr            0.10          0.3         0.10          100   
34      80    bpr            0.10          0.5         0.05          100   
35      80    bpr            0.10          0.5         0.05          150   

    precisions  
89    0.145649  
61    0.144860  
14    0.144663  
15    0.144663  
92    0.144072  
..         ...  
70    0.050378  
69    0.049294  
68    0.049294  
34    0.047915  
35    0.047915  

[96 rows x 7 columns]

In [41]:
model = LightFM(no_components=150,
                loss='bpr', 
                learning_rate=0.01, 
                item_alpha=0.1,
                user_alpha=0.1, 
                random_state=42,
                k=1,
                n=15,
                max_sampled=150)
model.fit((sparse_user_item > 0) * 1,  # user-item matrix из 0 и 1
          sample_weight=coo_matrix(user_item_matrix),
          user_features=csr_matrix(user_feat_lightfm.values).tocsr(),
          item_features=csr_matrix(item_feat_lightfm.values).tocsr(),
          epochs=10, 
          num_threads=20,
          verbose=True) 
users_ids_row = data_train_filtered['user_id'].apply(lambda x: userid_to_id[x]).values.astype(int)
items_ids_row = data_train_filtered['item_id'].apply(lambda x: itemid_to_id[x]).values.astype(int)

predictions = model.predict(user_ids=users_ids_row,
                            item_ids=items_ids_row,
                            user_features=csr_matrix(user_feat_lightfm.values).tocsr(),
                            item_features=csr_matrix(item_feat_lightfm.values).tocsr(),
                            num_threads=10)

data_train_filtered['score'] = predictions

predict_result = data_train_filtered[['user_id','item_id','score']][data_train_filtered.item_id != 999999].drop_duplicates().sort_values(by=['user_id','score'], ascending=False).groupby('user_id')['item_id']. \
            unique().reset_index()

df_result_for_metrics = result.merge(predict_result, on='user_id', how='inner')
precision = df_result_for_metrics.apply(lambda row: custom_precision(row['item_id'], row['actual'],k=5), axis=1).mean()


print(f"Precision: {precision}")

Epoch: 100%|████████████████████████████████████| 10/10 [00:20<00:00,  2.07s/it]


Precision: 0.1465353037766812


Лучшая точность Precision: ~0.15, параметры получились следующие:
no_components=150

loss='bpr'

learning_rate=0.01

item_alpha=0.1 - снижение параметра дало ощутимый результат

user_alpha=0.1

k=1 - как в предыдущем занятии, искать похожее на себя выгоднее.

n=15

max_sampled=150